content/blog/debugging-dde.md

```diff
@@ -0,0 +1,149 @@
+freed and deconstructed by then, so could rightly return undefined values.
+
+But this did expose a genuine difference between platforms in a variable; 
+narrowed down to the `while` loop above, which terminated early (or didn't 
+get entered at all), under certain 32-bit circumstances. Printing things within
+the loop - or sometimes before it - sometimes changed the behaviour and made
+things work, but it would be poor form to submit a PR that claims to fix a functional
+problem by printing a ton of waffle before-hand.
+
+My github commit titles around this time are telling. **Total Confusion Reigns** 
+towards **Wits' End** and later a  **Deeper Confusion** that's somehow worse than the 
+previous *total* one. Eventually, as hope ebbed away, the crucial summit arrived 
+at a commit aptly named **Another bizarre effort**.
+
+The question now moves from "why *doesn't* it work" to, "why *does* it":-
+
+```
+cond = (a>0) && (b>0);
+while (((a>0) && (b>0)) || cond) {
+  // a and b may get changed
+  cond = (a>0) && (b>0);
+}
+
+```
+The syntax is surely equivalent to the code at the top, if 
+more wordy with some redundency. A good PR might recommend
+cleaning it up back into the code at the top. Nevertheless,
+looking at `a` and `b` twice for each decision caused correct entry and exit
+of the loop, and the appveyor tests passed fully.
+
+## The Final Insult
+
+Good programmers should be cautious to blame their compiler, as in almost every
+case where suspicion falls on gcc, or Intel's or Microsoft's
+compiler, eventually analysis hunts down a memory leak or bad array write, or
+uninitialised variable; the compiler is shown to be correct and sane;
+the programmer a bit less so (probably on both points by that time).
+
+In this case though, gcc was to blame, and the optimised code even with `-O1` 
+changed the behaviour of the loop, breaking the simple test. It didn't optimise
+the loop away entirely, since all the many
+other tests worked consistently throughout the saga; it caused a subtle and rare
+change in behaviour - and only for 32-bits. If I had edited `%USERPROFILE%\Documents\.R\Makevars`
+and set
+```
+CFLAGS=-O0 -g -std=gnu99 -Wall
+```
+
+a bit earlier (which produces the right answer), I might have doubted gcc 
+sooner. Or would I? Faulty compiler optimisations and memory usage bugs can
+have similar side effects; both can do strange things when the code is tickled 
+by an innocuous print command. Both can disappear when you turn optimisation off. 
+(User memory bugs can also disappear only when you turn optimisation on...)
+
+We briefly looked at the assembler code produced by the compiler, (`-S` on the 
+gcc command line. It helps if the fragments of code you're comparing have
+exactly matching line numbers, since assembler code is made of gotos.) But 
+while we could tell the code was different with `-O1`, it would take a
+special kind of assembler programmer to show us why exactly the behaviour
+changed as it did.
+
+## Conclusion
```

Comment:
I think that another good takehome is that for platform-dependent code it is really useful to have CI set up on multiple platforms. This would have been been a nightmare to have picked up in user code in the wild

Reply:
Excellent point.